### PR TITLE
Bug 1810951 - Implement Glean-based crash pings in lib-crash

### PR DESCRIPTION
**Pull request by afranchuk**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1811928
https://bugzilla.mozilla.org/show_bug.cgi?id=1810951

### CODE REVIEW

**Comment by afranchuk** [comment]
This PR relies on #686 being merged first.

**Review by travis79 — APPROVED** [review]

**Comment by travis79** [comment]
This is great, one thing that may make this more robust in the future is some changes coming to Glean to make it easier to use outside of the main process on mobile: https://bugzilla.mozilla.org/show_bug.cgi?id=1807019

**Comment by afranchuk** [comment]
> This is great, one thing that may make this more robust in the future is some changes coming to Glean to make it easier to use outside of the main process on mobile: https://bugzilla.mozilla.org/show_bug.cgi?id=1807019

Cool! I'm not entirely sure if that would help here, since the main glean process may not exist at the time of crash, so it wouldn't be able to get the client info and/or know where to send the pings? But I admit I'm not intimately familiar with the process life cycle that's setup here.

File: android-components/components/lib/crash/src/main/java/mozilla/components/lib/crash/service/GleanCrashReporterService.kt
Comment (csadilek) [inline android-components/components/lib/crash/src/main/java/mozilla/components/lib/crash/service/GleanCrashReporterService.kt:193]: Any existing non-submitted crash files will cause this exception here, but probably not worth handling. We will just miss those crashes. That would happen when the app crashes and an update is applied to a version that has this change, before the app is restarted (and submits the crash). Then we couldn't parse the old file format anymore. That should be a (very) small % of clients though. Worth pointing out in case we see a drop while this rolls out.

File: android-components/components/lib/crash/src/test/java/mozilla/components/lib/crash/service/GleanCrashReporterServiceTest.kt
Comment (csadilek) [inline android-components/components/lib/crash/src/test/java/mozilla/components/lib/crash/service/GleanCrashReporterServiceTest.kt:22]: nit: Same not above, that this will fail the linter.

**Review by csadilek — COMMENTED** [review]
Thank you, this looks great! Made a first pass, and left some comments. Happy to approve once all reviews are in and the dependent PRs land.

I would like to revisit the decision to download the .yaml files, because it doesn't allow to build the app when offline anymore, and will likely cause some intermittent failures in CI too, but not a blocker.

File: android-components/components/lib/crash/build.gradle
Comment (csadilek) [reply android-components/components/lib/crash/build.gradle:131]: An option here is to add `compileOnly project(":browser-engine-gecko")` to https://github.com/mozilla-mobile/firefox-android/blob/5a961111addca08dcaa402d2a79b05920c04e731/android-components/components/lib/crash/build.gradle#L77. The comment actually claims it already does, which is outdated :).

Then it should be possible to get to this yaml file through the geckoview.aar? There would ideally be a separate step where we refactor this into a separate `lib-crash-glean` library, similar to `lib-crash-sentry`, but that's out of scope and can definitely be done later.

Comment (csadilek) [reply android-components/components/lib/crash/build.gradle:131]: Thinking about it more, I think we actually have to do something like the above here, because this currently fetches these files for a specific hard-coded Github tag, which can easily become outdated/incorrect and require manual updates on different branches. With the above, we would automatically use the right version of these .yaml files that match the Gecko version being used by the app (no matter what branch it is build from.)

Comment (csadilek) [reply android-components/components/lib/crash/build.gradle:131]: Another thought: If we're just going to fetch these files from a specific tag, and decide that's OK, we could also copy this files into the tree. We can keep them up-to-date with a Github action too until we move Fenix to M-C. This way we don't need to download them, as part of the build.

File: android-components/components/lib/crash/src/main/java/mozilla/components/lib/crash/service/GleanCrashReporterService.kt
Comment (afranchuk) [reply android-components/components/lib/crash/src/main/java/mozilla/components/lib/crash/service/GleanCrashReporterService.kt:26]: I didn't even want the wildcard, that was just the default "add this import" behavior that the IDE did :( I hate those features and usually turn them off, but didn't look through the settings to find it. I'll fix it.

File: android-components/components/lib/crash/build.gradle
Comment (afranchuk) [reply android-components/components/lib/crash/build.gradle:65]: I did consider the builtin android Json utilities, and also looked (in vain) for other "batteries included" options (aside from `kotlinx`). Like you pointed out, it would be extra work to write serialization boilerplate, and I wanted to leave it as easy as possible to add new data in the future.

File: android-components/components/lib/crash/build.gradle
Comment (afranchuk) [reply android-components/components/lib/crash/build.gradle:131]: > The comment actually claims it already does, which is outdated :).
Haha, I was wondering about that, thanks for validating my concern.

As far as the tag/version of the files being outdated: my reasoning was that it shouldn't actually _matter_ if they are outdated, because changes should only be additive in nature (as far as metrics are concerned) and without adding the new code to fill in new metrics, it wouldn't make a difference. But from a development perspective, it _would_ be nice if it always fetched the correct related version for a particular release. I'm not sure whether these files are actually included in the `geckoview.aar` at all though?

---

> Another thought: If we're just going to fetch these files from a specific tag, and decide that's OK, we could also copy these files into the tree. We can keep them up-to-date with a Github action too until we move Fenix to M-C. This way we don't need to download them, as part of the build.

We did consider just having the files in-tree (that would have made this simpler, obviously) but we were hesitant to have them diverge (though ultimately it would just be another update/programming task, and divergence in this way isn't much different than having the fixed tag).

I'm okay with having them in-tree. That seems like it would address both the offline-build issue (sorry about that, I didn't actually realize that offline builds were supported) and `probe-scraper` integration (the current approach requires a manual project be added). I guess `probe-scraper` would just have duplicates (under different project names) of the metrics/pings?

File: android-components/components/lib/crash/src/main/java/mozilla/components/lib/crash/service/GleanCrashReporterService.kt
Comment (afranchuk) [reply android-components/components/lib/crash/src/main/java/mozilla/components/lib/crash/service/GleanCrashReporterService.kt:193]: Yep, I acknowledge that. I didn't think it would be that high a priority (due to low likelihood) to make sure both formats are supported.

File: android-components/components/lib/crash/build.gradle
Comment (afranchuk) [reply android-components/components/lib/crash/build.gradle:131]: If they were in-tree, the only potential issue is that one platform (desktop, android, etc) is updated but other platforms are forgotten to be updated. But I think that's more-or-less a non-issue, since development to fill out new metrics will be needed anyway for each platform (and _maybe_ there will be platform-specific metrics).

We wanted the metrics to be the same for purposes of querying and joining ping data easily, however I think that's probably not that much of a concern with separate metrics definitions if we just make sure to use the same metrics names when applicable? I'm speaking somewhat naively here though, as I've not personally queried Glean ping data, and I don't know whether e.g. it's easy to join queries across project/application labels (or if that doesn't matter whatsoever).

File: android-components/components/lib/crash/src/main/java/mozilla/components/lib/crash/service/GleanCrashReporterService.kt
Comment (csadilek) [reply android-components/components/lib/crash/src/main/java/mozilla/components/lib/crash/service/GleanCrashReporterService.kt:26]: Thanks, same!

File: android-components/components/lib/crash/src/main/java/mozilla/components/lib/crash/service/GleanCrashReporterService.kt
Comment (csadilek) [reply android-components/components/lib/crash/src/main/java/mozilla/components/lib/crash/service/GleanCrashReporterService.kt:193]: Yeah, I agree we can live with that. Mostly commenting for documentation purposes.

File: android-components/components/lib/crash/build.gradle
Comment (csadilek) [reply android-components/components/lib/crash/build.gradle:131]: >  I'm not sure whether these files are actually included in the geckoview.aar at all though?

They are not currently, but we could technically add them. However, I think we're both leaning towards just adding the .yaml files to the tree now to prevent the issues above. The plan is for this repo to migrate to M-C later this year, at which point we can remove the duplication, and in the meantime the duplication isn't any worse than fetching a specific git tag. In both cases, a change in this repo needs to land to keep the files in sync with any changes happening on M-C. We either have to change the git tag or update the files directly. So that difference doesn't matter, it seems, plus then we can still build while offline, and avoid intermittent CI issues too.

If the files in M-C change a lot, we can hook up a Github action to open a PR and keep the files in sync. We have a few of those already.

Does that sound OK? @afranchuk @badboy @travis79 

@afranchuk Two minor points about landing this: 
- I think we can land these changes in one PR when ready. 
- We have an early code freeze (next week) for this cycle, because we're moving Fenix into this repo, as part of the Monorepo project and preparation. So, we should land this change on Feb. 14, first day of the 112 cycle. This also makes sure that the new metrics start at the beginning of 112, which is a nice consequence.

Comment (csadilek) [reply android-components/components/lib/crash/build.gradle:131]: @afranchuk The build is actually [failing](https://github.com/mozilla-mobile/firefox-android/pull/671/checks?check_run_id=11045687489) now, which is unexpected, given that the change to fetch .yaml files is part of this PR and Glean code generation should've happened?

```
[task 2023-02-01T21:18:10.581Z] e: /builds/worker/checkouts/vcs/android-components/components/lib/crash/src/main/java/mozilla/components/lib/crash/service/GleanCrashReporterService.kt: (20, 50): Unresolved reference: Pings
[task 2023-02-01T21:18:10.582Z] e: /builds/worker/checkouts/vcs/android-components/components/lib/crash/src/main/java/mozilla/components/lib/crash/service/GleanCrashReporterService.kt: (27, 50): Unresolved reference: Crash
[task 2023-02-01T21:18:10.583Z] e: /builds/worker/checkouts/vcs/android-components/components/lib/crash/src/main/java/mozilla/components/lib/crash/service/GleanCrashReporterService.kt: (91, 25): Unresolved reference: Pings
...
```

Comment (afranchuk) [reply android-components/components/lib/crash/build.gradle:131]: Hmm, that is unusual. I haven't seen any failures with what I've written (from various states of change wrt build state). Changing to local files will likely smooth that out.

Comment (afranchuk) [reply android-components/components/lib/crash/build.gradle:131]: Weirdest thing: I just changed my local copy to have copied definitions and removed all the downloading stuff, and now I'm seeing those same unresolved reference errors.

Comment (afranchuk) [reply android-components/components/lib/crash/build.gradle:131]: Nevermind, my rebase decided to stack rather than replace a commit (so it still had downloaded files). With the local files it seems okay. So I guess there was some way to get it into a state where it wouldn't generate the downloaded files. I would have thought the task dependencies were set up to avoid that :shrug:

File: android-components/components/lib/crash/metrics.yaml
Comment (rocketsroger) [inline android-components/components/lib/crash/metrics.yaml:62]: Since these never expires, please add the notification_emails: (Both android-probes@mozilla.com and the product manager's email).  This applies to below as well.

**Review by rocketsroger — COMMENTED** [review]
Looks good to me, ~~please start a data review as well~~. Thanks (Also didn't see that the data review was done in buzilla)

File: android-components/components/lib/crash/metrics.yaml
Comment (rocketsroger) [reply android-components/components/lib/crash/metrics.yaml:62]: Ignore my comment, didn't notice the emails were added above the `bugs:` section.

Comment (travis79) [reply android-components/components/lib/crash/metrics.yaml:62]: `expires: never` still requires an individual contributor to be responsible for the collection, and their email noted here.

File: android-components/components/lib/crash/build.gradle
Comment (csadilek) [reply android-components/components/lib/crash/build.gradle]: Ok, nice! The PR is green now too, and I like how this simplifies the build. From my perspective, this is all good now, but would like @badboy and @travis79 to check our conversation and conclusion here too :).

@afranchuk do you still want to keep that other [PR](https://github.com/mozilla-mobile/firefox-android/pull/686) separate? I think including it all here is fine.

Comment (travis79) [reply android-components/components/lib/crash/build.gradle]: After reading the conversation here, if keeping a separate copy of the files is no more or less of a pain than including them in the geckoview aar file and pulling them out, and makes offline building easier, I'd say that is fine. Since these metrics are organized by the application they are sent from, there shouldn't be any problem having the same names for metrics in different files/apps as long as the probe-scraper is configured correctly.

**Comment by badboy** [comment]
Please hold back on merging this, until I can check this against probe-scraper and to ensure this all works as expected. I'll give an update tomorrow.

**Comment by afranchuk** [comment]
@badboy no worries, it won't be merged until after the next release (https://github.com/mozilla-mobile/firefox-android/pull/671#discussion_r1096020575).

File: android-components/components/lib/crash/build.gradle
Comment (afranchuk) [reply android-components/components/lib/crash/build.gradle]: @csadilek I can remove the other PR. I only added it because I was unsure of what process should be followed WRT traceability and linking in bugzilla.

**Comment by badboy** [comment]
Ok, this is a bit weird because our tooling might complain about duplicated metrics. I opened [a bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1815432) to figure out what we need to do. This is currently in the data team's hand, no action from your side to clear this up.

**Comment by afranchuk** [comment]
Okay. That's what was unclear to me: whether separate glean projects/repositories can have the same named metrics. I assumed so only _because_ of the existence of the project/library/application distinction in the probe-scraper, but perhaps that's just used for traceability?

**Comment by afranchuk** [comment]
If we must, we can rename the top-level category. But hopefully that's not necessary!

**Review by csadilek — APPROVED** [review]

**Comment by csadilek** [comment]
Adding the "Do not land label" to wait for next release and for [bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1815432).

File: android-components/components/lib/crash/build.gradle
Comment (csadilek) [reply android-components/components/lib/crash/build.gradle]: @afranchuk OK, thanks, I will close the other PR and link it to this one then.

**Comment by badboy** [comment]
> If we must, we can rename the top-level category. But hopefully that's not necessary!

Yeah, the whole point is that we keep the names the same across the app. I'm trying to make that happen.

**Comment by afranchuk** [comment]
Awesome, I'll wait and see how that bug plays out.

**Comment by afranchuk** [comment]
[bug 1815432](https://bugzilla.mozilla.org/show_bug.cgi?id=1815432) has been moved along (with the changes to probe-scraper merged today), so we should be free to merge this soon, if not now. I'm not sure whether the solution to the bug involves this being merged too, but I think it does as that will allow probe-scraper to get the new metrics.

**Comment by csadilek** [comment]
@afranchuk feel free to land when you're ready using the "needs landing" label. Our preference is not to land bigger new changes on Fridays as those will otherwise roll out over the weekend, but this seems pretty safe. Otherwise, I would recommend just landing Monday morning.

**Comment by afranchuk** [comment]
I will double-check with the Glean folks and then set to land on Monday.

**Comment by afranchuk** [comment]
It seems like CI is trying to use kotlin 1.8.10? But the repository dependencies are set up for kotlin 1.7.20? That's what (I believe) is causing the failures (in downloading specific versions of ksp).

**Comment by csadilek** [comment]
> It seems like CI is trying to use kotlin 1.8.10? But the repository dependencies are set up for kotlin 1.7.20? 

@afranchuk ah no, sorry, Kotlin was upgraded to 1.8.10 today and Mergify rebased your PR: https://github.com/mozilla-mobile/firefox-android/pull/1099. I think we will need to use ksp 1.0.9 now (instead of 1.0.8) for Kotlin 1.8.10, looking at: https://github.com/google/ksp/releases/tag/1.8.10-1.0.9

**Comment by afranchuk** [comment]
Doh, I will rebase it. The way they release ksp as a combination version (but without defined supported version windows) is kinda confusing. Maybe figuring out how to make things work better kapt work better with glean should be investigated (though it seems like ksp is the future).

**Comment by csadilek** [comment]
Agreed, versioning is confusing, also had to look twice. Hopefully it all works for now if you change ksp to 1.0.9 here. 🤞

**Pull request merged**
